src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,411 @@
+    private long solvingStart;
+    private long solvingEnd;
+
+    @Override
+    public InferenceSolution solve(Map<String, String> configuration, Collection<Slot> slots,
+            Collection<Constraint> constraints, QualifierHierarchy qualHierarchy,
+            ProcessingEnvironment processingEnvironment) {
+
+        InferenceSolution solution = null;
+
+        configure(configuration);
+        configureLattice(qualHierarchy);
+        Serializer<?, ?> defaultSerializer = createSerializer(backEndType, lattice);
+
+        if (useGraph) {
+            this.graphBuildingStart = System.currentTimeMillis();
+            this.constraintGraph = generateGraph(slots, constraints, processingEnvironment);
+            this.graphBuildingEnd = System.currentTimeMillis();
+            StatisticRecorder.record(StatisticKey.GRAPH_GENERATION_TIME, (graphBuildingEnd - graphBuildingStart));
+            solution = graphSolve(constraintGraph, configuration, slots, constraints, qualHierarchy,
+                    processingEnvironment, defaultSerializer);
+        } else {
+            realBackEnd = createBackEnd(backEndType, configuration, slots, constraints, qualHierarchy,
+                    processingEnvironment, lattice, defaultSerializer);
+            solution = solve();
+        }
+
+        if (solution == null) {
+            ErrorReporter.errorAbort("Null solution detected!");
+        }
+
+        if (collectStatistic) {
+            Map<String, Integer> modelRecord = recordSlotConstraintSize(slots, constraints);
+            PrintUtils.printStatistic(StatisticRecorder.getStatistic(), modelRecord, backEndType,
+                    useGraph, solveInParallel);
+            PrintUtils.writeStatistic(StatisticRecorder.getStatistic(), modelRecord, backEndType,
+                    useGraph, solveInParallel);
+        }
+        return solution;
+    }
+
+    /**
+     * This method configures following arguments: backEndType, useGraph,
+     * solveInParallel, and collectStatistic
+     * 
+     * @param configuration
+     */
+    private void configure(final Map<String, String> configuration) {
+
+        final String backEndName = configuration.get(Constants.BACK_END_TYPE);
+        final String useGraph = configuration.get(Constants.USE_GRAPH);
+        final String solveInParallel = configuration.get(Constants.SOLVE_IN_PARALLEL);
+        final String collectStatistic = configuration.get(Constants.COLLECT_STATISTIC);
+
+        if (backEndName == null || backEndName.equals("MaxSAT")) {
+            // Configure back end type. Default is MaxSAT.
+            this.backEndType = Constants.MAX_SAT;
+        } else if (backEndName.equals("Lingeling")) {
+            this.backEndType = Constants.LINGELING;
+        } else if (backEndName.equals("LogiQL")) {
+            this.backEndType = Constants.LOGIQL;
+        } else {
+            ErrorReporter.errorAbort("Back end \"" + backEndName + "\" has not been implemented yet.");
```

Comment:
We do want to add new back ends into the framework, but is there a way to control back end type without the predefined set? Users choose back end type from command line, and unless they type full qualified name of the back end, we don't know where to find the back end without the predefined set. Is this right?